akka-http2-support/src/main/scala/akka/http/impl/engine/http2/RequestParsing.scala

```diff
@@ -35,72 +39,112 @@ private[http2] object RequestParsing {
+          }
           if (remoteAddressHeader.isDefined) headers += remoteAddressHeader.get
 
           val entity =
             if (subStream.data == Source.empty || contentLength == 0) HttpEntity.Empty
             else if (contentLength > 0) HttpEntity.Default(contentType, contentLength, subStream.data)
             else HttpEntity.Chunked.fromData(contentType, subStream.data)
 
-          val uri = Uri(scheme, authority).withPath(Uri.Path(path))
+          val (path, rawQueryString) = pathAndRawQuery
+          val uri = Uri(scheme, authority, path, rawQueryString)
           HttpRequest(
             method, uri, headers.result(), entity, HttpProtocols.`HTTP/2.0`
           )
         } else remainingHeaders.head match {
           case (":scheme", value) ⇒
-            rec(remainingHeaders.tail, method, value, authority, path, contentType, contentLength, headers)
+            checkUniquePseudoHeader(":scheme", scheme)
+            checkNoRegularHeadersBeforePseudoHeader(":scheme", seenRegularHeader)
+            rec(remainingHeaders.tail, method, value, authority, pathAndRawQuery, contentType, contentLength, cookies, seenRegularHeader, headers)
           case (":method", value) ⇒
+            checkUniquePseudoHeader(":method", method)
+            checkNoRegularHeadersBeforePseudoHeader(":method", seenRegularHeader)
             val m = HttpMethods.getForKey(value).getOrElse(malformedRequest(s"Unknown HTTP method: '$value'"))
-            rec(remainingHeaders.tail, m, scheme, authority, path, contentType, contentLength, headers)
-          case (":path", path) ⇒
-            rec(remainingHeaders.tail, method, scheme, authority, path, contentType, contentLength, headers)
+            rec(remainingHeaders.tail, m, scheme, authority, pathAndRawQuery, contentType, contentLength, cookies, seenRegularHeader, headers)
+          case (":path", value) ⇒
+            checkUniquePseudoHeader(":path", pathAndRawQuery)
+            checkNoRegularHeadersBeforePseudoHeader(":path", seenRegularHeader)
+            // QUESTION: Should this logic be exposed in a method in either Uri or Query? (would be more consistent)
+            // QUESTION: Should we catch the parsing exception and throw a more specific type?
+            val newPathAndRawQuery = new UriParser(new StringBasedParserInput(value), UTF8, serverSettings.parserSettings.uriParsingMode).parseHttp2PathPseudoHeader()
+            rec(remainingHeaders.tail, method, scheme, authority, newPathAndRawQuery, contentType, contentLength, cookies, seenRegularHeader, headers)
           case (":authority", value) ⇒
-            val authority = Uri.Authority.parse(value)
-            rec(remainingHeaders.tail, method, scheme, authority, path, contentType, contentLength, headers)
+            checkNoRegularHeadersBeforePseudoHeader(":authority", seenRegularHeader)
+            // QUESTION: Should we treat this request as malformed if there are duplicate :authority headers? (The spec says nothing)
```

Comment:
It says

>  All HTTP/2 requests MUST include exactly one valid value for the
   ":method", ":scheme", and ":path" pseudo-header fields,

I guess the reason it doesn't say so for the `:authority` field is because it might be missing but I don't think we need to allow multiple instances.